packages/insured-bridge-relayer/src/index.ts

```diff
@@ -52,8 +53,6 @@ export async function run(logger: winston.Logger, l1Web3: Web3): Promise<void> {
 
     // Construct a web3 instance running on L2.
     const l2Web3 = getWeb3ByChainId(config.activatedChainIds[0]);
-    const latestL2BlockNumber = await l2Web3.eth.getBlockNumber();
```

Comment:
NOTE THIS CHANGE! @nicholaspai I made this as we need the ability to see `TokensBridged` events all the way back, past the `l2BlockLookback`. This means that this implementation _only_ works on Alchemy and NOT infura as we need to look back past what that provider can support. to address this we should update the L2client to facilitate pagination to work around this restriction.

let me know what you think!

Reply:
Hm I don't like that this change would force us to no longer work with Infura. Why don't you separate event querying between two flows:
- the usual getPastEvents call that goes from `{fromBlock: l2StartBlock, toBlock: latest}` for events like `FundsDeposited` 
- Add a second event querying path that paginates from `{fromBlock: _fromBlock, toBlock: _toBlock}`, and sits inside a `for` loop that keeps decrementing `_fromBlock` until `_fromBlock <= 0` and sets `_toBlock = _fromBlock`.

I think pagination is the way to go instead of cuttting off compatibility with Infura

Reply:
Given that it needs a pretty different lookback and it's doing a different query, do you think it makes sense to do this logic somewhere outside of this client (maybe a separate client or maybe in the finalizer logic itself)?

Agree that pagination (and adding a utility for it!) seems like a good idea here. As long as we know how many blocks we're trying to go back, we could also do this pagination in parallel (with some limit to the number of in-flight requests using bluebird, maybe?) so we don't have to worry about this bot taking a long time to run due to the serialized queries.

Going back to 0 might not be ideal, though, especially if we're doing pagination.

Reply:
@nicholaspai agreed that pagination is the solution. the way this PR is structured right now, however, makes it not work with infura as we currently don't have the pagination logic. I also agree having two flows for different kinds of events seems useful. ideally though I'd like to create a `paginationEventLookup` or something equivalent that will abstract away this from the user so it does not have to sit in the client.

@mrice32 putting the query somewhere else does make sense! this is how we had structured things originally (the finalizer itself fetched the whitelist event data). we ended up rolling this into the client when we needed access to the same data in another location. I agree though in this context we wont really need the data anywhere else and pulling it in for all bots (even those that don't run the finalizer) is a bit of a waist on resources. Regarding how far we go back we don't need to go all the way to block 0, we can simply go back to `bridgeDepositBoxDeployData` from the config file https://github.com/UMAprotocol/protocol/blob/88bd694ad916cb3ee5e97102e27d073b1b9fdf67/packages/insured-bridge-relayer/src/RelayerConfig.ts?_pjax=%23js-repo-pjax-container%3Afirst-of-type%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%3Afirst-of-type%2C%20%5Bdata-pjax-container%5D%3Afirst-of-type#L32

Reply:
After looking more at this I think the best solution is to go with what @mrice32 described by pulling the event query into the `CrossDomainFinalizer`. this has a number of advantages, least of which we don't need to fetch any of this data unless you want to run this particular bot. this makes "normal" bot operation less taxing on nodes and user resources. 

while considering this, I also realized that we capute a fair amount of unnecessary data in the `L2Client`. in particular, we fetch `WhitelistToken` and store it within the client. this is only used with the method `isWhitelistedToken` (see [here](https://github.com/UMAprotocol/protocol/blob/05930397cd67e018bc09d6c5d8169514b365a4ee/packages/financial-templates-lib/src/clients/InsuredBridgeL2Client.ts?_pjax=%23js-repo-pjax-container%3Afirst-of-type%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%3Afirst-of-type%2C%20%5Bdata-pjax-container%5D%3Afirst-of-type#L59)) but we never use this data anywhere in the bots. Moreover, the current implementation of this method will _not_ contain all expected tokens as the lookback, as injected into this client currently in master, prevents the L2 client from looking back much past ~100k events!

I've refactored this to do the following:
1) I removed the `isWhitelistedToken` and associated event retrieval and data from the L2Client, thereby making it lighter and easier to work with.
2) I refactored the fetching logic in the L2 client, previously called `getFundsDepositedEvents` to `getBridgeDepositBoxEvents` to make things more modular and clean.
3) I pulled the logic for fetching `TokensBridged` events _directly_ into the `CrossDomainFinalizer`. while this is a bit convoluted in that we fetch this data in a location that is _not_ a "client", it does make the bots lighter (only fetch the data if you run this module) and solves where the pagination logic should be.

Reply:
I worked on creating a basic pagination block searcher. Logic looked basically like this(rough example from testing):
```ts
const currentBlock = await web3.eth.getBlockNumber();
  const firstBlock = 2811998;
  const maxBlockChunk = 99900;

  const paginationChunks = Math.ceil((currentBlock - firstBlock) / maxBlockChunk);

  let blockSearchIndecides = [];
  for (let i = 0; i < paginationChunks; i++) {
    blockSearchIndecides[i] = {
      fromBlock: firstBlock + maxBlockChunk * i + (i == 0 ? 0 : 1),
      toBlock: firstBlock + maxBlockChunk * (i + 1),
    };
    if (blockSearchIndecides[i].toBlock > currentBlock)
      blockSearchIndecides[i].toBlock = currentBlock;
  }

  const tokensBridgedEvents = await Promise.all(
    blockSearchIndecides.map((searchSettings) =>
      depositBox.getPastEvents('TokensBridged', searchSettings)
    )
  );
```

Issue is that this does not play nicely with our RPC endpoints! when run against infura this blows up the node and you get back an invalid JSON RPC response. I added `BluebirdPromise.all` to try and set maximum `concurrency` settings but I can only run maximum 1 concurrency at the same time on infura or this errors out :/ From when the contracts were deployed, for a chunk size of 99900 we need to make 10 sequential queries to build up the full event set. This takes almost 40 seconds.


for now, I think the solution should be to not do any pagination and simply run this particular module on `alchemy` only, thereby avoiding this problem altogether. Infura in the future will support longer look-backs on Arbitrum and therefore I don't see any particular point in fighting this restriction any further seeing pagination does not work.